### PR TITLE
Add ZChannel.map

### PR DESCRIPTION
**Pull request by sideeffffect**
* add `ZChannel.map`
* implement `mapOut` and `identity` in terms of it
* use in `ZPipeline.map`

### CODE REVIEW

File: streams/shared/src/main/scala/zio/stream/ZChannel.scala [hunk 1]
```diff
@@ -1345,6 +1325,15 @@ object ZChannel {
       releaseMap.releaseAll(exit, ExecutionStrategy.Sequential)
     }.mapOut(_._1)
 
+  def map[Err, InElem, OutElem, Done](
```
Comment (adamgfraser) [inline streams/shared/src/main/scala/zio/stream/ZChannel.scala:1328]: `mapOut`

File: streams/shared/src/main/scala/zio/stream/ZChannel.scala [hunk 2]
```diff
@@ -1345,6 +1325,15 @@ object ZChannel {
...
+    f: InElem => OutElem
+  )(implicit trace: ZTraceElement): ZChannel[Any, Err, InElem, Done, Err, OutElem, Done] =
+    readWith(
+      (in: InElem) => write(f(in)) *> map[Err, InElem, OutElem, Done](f),
```
Comment (adamgfraser) [inline streams/shared/src/main/scala/zio/stream/ZChannel.scala:1332]: Type inference is pretty terrible here.

**Review by adamgfraser — COMMENTED** [review]
There are a lot of `ZChannel` operators that could be written as channel constructors. Can definitely see them being useful in some situations but does create additional ways to do things and type inference is generally going to be worse. You can also always construct a channel like this by doing `ZChannel.identity.mapOut(f)`.

**Comment by sideeffffect** [comment]
Ah, yes, I see. In that case this PR doesn't bring much to the table.
If you see anything worth salvaging, please let me know, otherwise I'll just close it :+1:

**Pull request closed without merging**